processing/src/test/java/org/apache/druid/segment/join/JoinablesTest.java

```diff
@@ -190,6 +196,176 @@ public boolean isDirectlyJoinable(DataSource dataSource)
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_1)).anyTimes();
+    EasyMock.replay(analysis);
+
+    Optional<byte[]> cacheKey_1 = joinables.computeJoinDataSourceCacheKey(analysis);
+    Assert.assertTrue(cacheKey_1.isPresent());
+    Assert.assertNotEquals(0, cacheKey_1.get().length);
+
+    PreJoinableClause clause_2 = makeGlobalPreJoinableClause("dataSource_1", "x == \"h.x\"", "h.");
+    EasyMock.reset(analysis);
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_2)).anyTimes();
+    EasyMock.replay(analysis);
+    Optional<byte[]> cacheKey_2 = joinables.computeJoinDataSourceCacheKey(analysis);
+
+    Assert.assertNotEquals(cacheKey_1, cacheKey_2);
+  }
+
+  @Test
+  public void test_computeJoinDataSourceCacheKey_keyChangesWithJoinable()
+  {
+    DataSourceAnalysis analysis = EasyMock.mock(DataSourceAnalysis.class);
+    Joinables joinables = new Joinables(new JoinableFactoryWithCacheKey());
+
+    PreJoinableClause clause_1 = makeGlobalPreJoinableClause("dataSource_1", "x == \"j.x\"", "j.");
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_1)).anyTimes();
+    EasyMock.replay(analysis);
+
+    Optional<byte[]> cacheKey_1 = joinables.computeJoinDataSourceCacheKey(analysis);
+    Assert.assertTrue(cacheKey_1.isPresent());
+    Assert.assertNotEquals(0, cacheKey_1.get().length);
+
+    PreJoinableClause clause_2 = makeGlobalPreJoinableClause("dataSource_2", "x == \"j.x\"", "j.");
+    EasyMock.reset(analysis);
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_2)).anyTimes();
+    EasyMock.replay(analysis);
+    Optional<byte[]> cacheKey_2 = joinables.computeJoinDataSourceCacheKey(analysis);
+
+    Assert.assertNotEquals(cacheKey_1, cacheKey_2);
+  }
+
+  @Test
+  public void test_computeJoinDataSourceCacheKey_sameKeyForSameJoin()
+  {
+    DataSourceAnalysis analysis = EasyMock.mock(DataSourceAnalysis.class);
+    Joinables joinables = new Joinables(new JoinableFactoryWithCacheKey());
+
+    PreJoinableClause clause_1 = makeGlobalPreJoinableClause("dataSource_1", "x == \"j.x\"", "j.");
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_1)).anyTimes();
+    EasyMock.replay(analysis);
+
+    Optional<byte[]> cacheKey_1 = joinables.computeJoinDataSourceCacheKey(analysis);
+    Assert.assertTrue(cacheKey_1.isPresent());
+    Assert.assertNotEquals(0, cacheKey_1.get().length);
+
+    PreJoinableClause clause_2 = makeGlobalPreJoinableClause("dataSource_1", "x == \"j.x\"", "j.");
+    EasyMock.reset(analysis);
+    EasyMock.expect(analysis.getPreJoinableClauses()).andReturn(Collections.singletonList(clause_2)).anyTimes();
+    EasyMock.replay(analysis);
+    Optional<byte[]> cacheKey_2 = joinables.computeJoinDataSourceCacheKey(analysis);
+
+    Assert.assertNotEquals(cacheKey_1, cacheKey_2);
```

Comment:
Should this be `Assert.assertArrayEquals(cacheKey_1.get(), cacheKey_2.get());`?

Reply:
Similar comment for other tests you added. Should they use `Assert.assertFalse(Arrays.equals(cacheKey_1.get(), cacheKey_2.get()));`?

Reply:
oops. Thanks for catching this.